File: node_engine/src/checks/DeviceDependentEventHandlerCheck.js
const { getLineNumber, containsAttribute } = require("./TextUtilities");

const elementRegex = /<(?<tag>[a-z0-9]+)(?<attrs>[^>]*)>/gi;
const mouseClickEvents = ["onclick", "onmousedown", "onmouseup", "ondblclick"];
const mouseHoverEvents = ["onmouseover", "onmouseout", "onmouseenter", "onmouseleave"];
const keyboardEvents = ["onkeydown", "onkeypress", "onkeyup"];
const focusEvents = ["onfocus", "onblur"];

const containsAny = (attributes, names) => names.some((name) => containsAttribute(attributes, name));

const DeviceDependentEventHandlerCheck = {
  id: "device-dependent-event-handler",
  applicableKinds: ["html", "htm", "cshtml", "razor"],
  run(context, rule) {
    const issues = [];

    for (const match of context.content.matchAll(elementRegex)) {
      const attrs = match.groups?.attrs ?? "";
      const hasMouseClick = containsAny(attrs, mouseClickEvents);
      const hasMouseHover = containsAny(attrs, mouseHoverEvents);
      if (!hasMouseClick && !hasMouseHover) {
        continue;
      }

      const hasKeyboard = containsAny(attrs, keyboardEvents);
      const hasFocus = containsAny(attrs, focusEvents);
      if ((hasMouseClick && !hasKeyboard) || (hasMouseHover && !hasFocus)) {
        const line = getLineNumber(context.content, match.index);
        issues.push({
          ruleId: rule.id,
          checkId: DeviceDependentEventHandlerCheck.id,
          filePath: context.filePath,
          line,
          message: "Mouse-specific event handler lacks keyboard equivalent.",
          evidence: match[0]
        });
      }
    }

    return issues;
  }
};

module.exports = { DeviceDependentEventHandlerCheck, containsAny };

File: src/Scanner.Core/Checks/DeviceDependentEventHandlerCheck.cs
using System.Text.RegularExpressions;
using System.Linq;
using Scanner.Core.Rules;

namespace Scanner.Core.Checks;

/// <summary>
/// Checks for mouse-specific event handlers without keyboard equivalents.
/// </summary>
public sealed class DeviceDependentEventHandlerCheck : ICheck
{
    private static readonly Regex ElementRegex = new("<(?<tag>[a-z0-9]+)(?<attrs>[^>]*)>", RegexOptions.IgnoreCase | RegexOptions.Compiled);
    private static readonly string[] MouseClickEvents = { "onclick", "onmousedown", "onmouseup", "ondblclick" };
    private static readonly string[] MouseHoverEvents = { "onmouseover", "onmouseout", "onmouseenter", "onmouseleave" };
    private static readonly string[] KeyboardEvents = { "onkeydown", "onkeypress", "onkeyup" };
    private static readonly string[] FocusEvents = { "onfocus", "onblur" };

    /// <inheritdoc />
    public string Id => "device-dependent-event-handler";

    /// <inheritdoc />
    public IReadOnlyCollection<string> ApplicableKinds { get; } = new[] { "html", "htm", "cshtml", "razor" };

    /// <inheritdoc />
    public IEnumerable<Issue> Run(CheckContext context, RuleDefinition rule)
    {
        var issues = new List<Issue>();

        foreach (Match match in ElementRegex.Matches(context.Content))
        {
            var attrs = match.Groups["attrs"].Value;
            var hasMouseClick = ContainsAny(attrs, MouseClickEvents);
            var hasMouseHover = ContainsAny(attrs, MouseHoverEvents);
            if (!hasMouseClick && !hasMouseHover)
            {
                continue;
            }

            var hasKeyboard = ContainsAny(attrs, KeyboardEvents);
            var hasFocus = ContainsAny(attrs, FocusEvents);
            if ((hasMouseClick && !hasKeyboard) || (hasMouseHover && !hasFocus))
            {
                var line = TextUtilities.GetLineNumber(context.Content, match.Index);
                issues.Add(new Issue(rule.Id, Id, context.FilePath, line, "Mouse-specific event handler lacks keyboard equivalent.", match.Value));
            }
        }

        return issues;
    }

    private static bool ContainsAny(string attributes, IEnumerable<string> names)
        => names.Any(name => TextUtilities.ContainsAttribute(attributes, name));
}
